### ISSUE TITLE
Possible race condition with disconnect request

### ISSUE THREAD

**Issue by TylerLeonhardt** [microsoft/debug-adapter-protocol#126]
Maybe I have this wrong but for my statement:
* debuggee will be my "debug adapter server" (aka the process separate to vscode handling DAP messages) (it's the DAP equivalent to a Language Server)
* debug adapter is in a vscode extension (aka the client) - I have a [NamedPipeDebugAdapter](https://github.com/PowerShell/vscode-powershell/blob/master/src/debugAdapter.ts#L9).

The `disconnect` request is a very odd request to me because it not only tells the debuggee to terminate (if it was a launch request), but it also requires a response back to the debug adapter.

It's very easy to fall into a situation where there's a race condition between responding to the message and the debuggee terminating itself.

If the debuggee terminates itself first, then it never has a chance to respond to the `disconnect` message that the client sent... and so vscode shows an error toast saying "timeout 1000ms passed".

If the debuggee responds to the message first, and I'm using a named pipe or a tcp port as the communication layer, there's a possible race condition where launching the debuggee again attaches to the port or pipe that's in the process of being closed rather than a newly created process.

Do you have any guidance here? My biggest wish would be that the disconnect request should really be 'fire and forget' from the vscode side to prevent DAP implementors from hitting this race condition.

**Comment by TylerLeonhardt** [comment]
I think it was made clear to me by @weinand that my definitions are wrong...

In my scenario, the debug adapter (aka the thing handling DAP messages) is in the same process as the debuggee.

**Comment by weinand** [comment]
@TylerLeonhardt thanks for creating this issue.

Just some terminology so that we are on the same page as the DAP spec:
- the target of debugging (typically the end user's program) is the "debug target" or the "debuggee" .
- the code that handles DAP messages and talks to a real debugger or runtime is the "debug adapter" or "DA" (because it "adapts" the DAP protocol to a real debugger or runtime but is not itself a debugger).
- a debug adapter can be implemented in different ways:
  - as a standalone program (called "debug adapter") that handles a single debug session, 
  - as a standalone server (called "DAP server" ) that handles multiple debug sessions concurrently,
  - as an inline implementation that runs inside some other program or extension (your case).
  - etc.

In the original issue description from above "debuggee" is used for the debug adapter which is a bit confusing...

---

Now back to the "disconnect" issue.

The spec says:
> The ‘disconnect’ request is sent from the client to the debug adapter in order to stop debugging.
It asks the debug adapter to disconnect from the debuggee and to terminate the debug adapter.
If the debuggee has been started with the ‘launch’ request, the ‘disconnect’ request terminates the debuggee.
If the ‘attach’ request was used to connect to the debuggee, ‘disconnect’ does not terminate the debuggee.


So I read this as follows:

after the DA has received a ‘disconnect’ request from the client (e.g. VS Code), the DA disconnects from the debuggee.
If the DA had launched the debuggee initially, it "owns" it and it is responsible to terminate it.
If the DA was attached to the debuggee initially, it does not "own" it and so it must not terminate it.

The DA is free to decide when to send a response for the ‘disconnect’ request:
- it could send a response immediately before doing anything (this is the "fire and forget" approach)
- it could send a response immediately after disconnecting from the debuggee. 
- it could send a response after terminating the debuggee (only in the 'launch' case).

Then the DA terminates (after properly flushing the output channel so that the response could be successfully sent).
Terminating means different things for the different implementation options:
- a standalone debug adapter would just end or call "exit(...)"
- a DAP server would end the session by closing the comm channel.

---

So where exactly is the race condition?

**Comment by TylerLeonhardt** [comment]
In doing this exercise, I _think_ I've answered my own question, but here it is.

Consider the following architecture:

![image](https://user-images.githubusercontent.com/2644648/86816290-7f9e3e00-c038-11ea-8182-503a95a12cd1.png)

The server is spun up via other means (it's also a LSP server!) and creates a Named Pipe/UNIX Domain Socket that allows for a single connection (for security purposes and also because only one client can connect at a time) for the inline DA in vscode.

Now, when the inline DA sends a `disconnect` to the DA server side, it:
* attempts to send a `disconnect` response
* attempts to dispose of the `NamedPipeServerStream` that is the connection between the client and server

> NOTE: Right after this happens, the DA server side recreates the `NamedPipeServerStream` to reinitialize it to only allow a single connection

Now here's the race condition:

* What happens if we dispose of the `NamedPipeServerStream` before the `disconnect` response goes back?
    * Answer: VS Code shows a notification saying there was a timeout waiting for a response
* What happens if we send the `disconnect` response but don't dispose of the `NamedPipeServerStream` fast enough?
    * Answer: The next debug connection will connect to the old NamedPipe that is getting cleaned up and will cause things to be in a bad state

Ideally we:
* send the response to `disconnect` and then
* immediately clean everything up

But that's easier said than done in the [C# lib](https://github.com/Omnisharp/csharp-language-server-protocol) at this time

However, what I _might_ be able to do is:

Our DA server side shouldn't need to dispose of the named pipe and could potentially allow multiple connections over time (I have to see if this is configurable, otherwise I may have to handle the logic myself)

I have some security concerns with this approach, but it might be the right idea (and is similar to how TCP-based DA's work today, I believe: have something open the whole time rather than recreating on disconnect)

**Comment by polinasok** [comment]
@TylerLeonhardt @weinand

I came across this issue while doing research for our work to add native DAP support to the delve debugger for Go. And looking at the picture above, I think we are dealing with something similar. We do have to compile and launch the Go program as a separate process, so in our case the debugger and the user process would be two different boxes, but the debugger and the debug adapter would be sharing a box. And so there is the same question of the best way to connect VSCode to this Debugger/DA. Should we have a lean adapter in-between that is a DebugAdapterExecutable talking over stdin/out to vscode and over a socket to the debugger and just forwards the messages? Should we inline the forwarding adapter? Or should we skip it all-together and go for the model with the fewest moving parts using a standalone DAP server adapter with [DebugAdapterServer API](https://github.com/microsoft/vscode-mock-debug/blob/9174fca42890637ba5ed11ae18321f6f212d02c7/src/extension.ts#L80-L104)
```
VSCODE <=dap over socket=> DELVE DEBUGGER / SERVER / DEBUG ADAPTER <=ptrace=> DEBUGGEE
```
So my question based on the above picture is why do you use the inline DA to forward the messages over a socket instead of having vscode talk to the DA server side directly over that socket (e.g. with vscode.DebugAdapterNamedPipeServer)?

As for the disconnect race condition. We are considering having a long-running server that can accept a client connection,  initialize, launch/attach, do some debugging, then disconnect, then accept another client connection, rinse repeat. But for version 0, we were thinking of just single-use DA server that would terminate itself on disconnect. Let's say instead of picking a random port every time, we reuse a port. Is this what you would be concerned about?

```
Time 0
 | vscode requests disconnect
 | DA at :port responds to disconnect
 | DA at :port shuts everything down             vscode cleans up and turns off action bar
 |    ... still working ...                      user launches another debug session
 |    ... still working ...                      ext attempts to start new DA server at :port
 |    ... still working ...                      gets "address already in use" error
 | DA at :port is gone              
 |
Time N
```

I see how this can get tricky if the extension launches the debug server directly. But in your case you have another adapter on the other side of the pipe. Why does disposing of the NamedPipeServerStream result in a timeout waiting for the response? Does it take too long? Is there a way to reconfigure the timeout to allow more time for clean up?

>>>My biggest wish would be that the disconnect request should really be 'fire and forget' from the vscode side to prevent DAP implementors from hitting this race condition.

If disconnect request did not expect a response, wouldn't that behave the same as receiving an immediate response before everything is shut down in reality, so if the debug session is relaunched right away, you would end up trying to reuse something that is still being cleaned up? Seems like the only way to avoid that from the editor side is to have a mandatory delay before relaunching an adapter connection since the editor doesn't have access to whatever server we might be blocked on to check if things are ready to go again. And who knows if that will be long enough. Otherwise, we as implementors must do the checking ourselves or come up with a set up where nothing gets reused in conflicting ways (e.g. use long running server or use different port every time).

**Comment by puremourning** [comment]
This seems focussed on vscode not on DAP. I think the explanation of the protocol is clear. For the Delve case, this seems very similar to CodeLLDB (codelldb is the debugger and debug adapter and opens a tcp port for communication in single session mode).

**Comment by polinasok** [comment]
vscode sets a strong precedent for the implementors of DAP adapters because it is the most common client that they target. And then when that same adapter is reused by other editors, those are likely to try to mimic what vscode does to be fully in sync with the adapters. That's why the boundary between general DAP discussions and vscode discussions is often crossed.

The protocol spec is seemingly simple and comprehensive, but when you get to the nitty-gritty of the implementation, it is not always clear what to do because the devil i in the details, which are not covered by the spec.  So people turn to vscode for a real-life specification of what might be and what to work against. If that is to be avoided, the protocol documentation needs more details, both on what is expected and not expected from the adapter and what is expected and not expected from the client on the other end.

For example, in this particular case, if you think of a debug session as a single independent entity, then it doesn't really matter when you send back the disconnect response as long as you send it back as per the spec. But in real life, debug sessions happen back-to-back, so that boundary between the two sessions becomes a real implementation consideration. And that is not covered in the protocol. Not only it is not clear if the editor or the adapterare to safeguard against such conflicts, it is not even surfaced as a consideration for either party to have in mind when implementing the protocol. The spec could say something like: "Upon receiving the disconnect response, the editor is free to clean up. For example, it might update the UI to enable the next debug session. The implementors of the adapter should therefore make sure that any resources shared between the adapter invocations are cleared in time." And it that case, we would not need to reference vscode as a de-facto spec.

**Comment by TylerLeonhardt** [comment]
@polinasok 

> So my question based on the above picture is why do you use the inline DA to forward the messages over a socket instead of having vscode talk to the DA server side directly over that socket (e.g. with vscode.DebugAdapterNamedPipeServer)?

It didn't exist at the time. I actually contributed `vscode.DebugAdapterNamedPipeServer` to vscode... but never got around to moving vscode-powershell to use it. However, a similar class was made before I made the contribution:
https://github.com/PowerShell/vscode-powershell/blob/28e6a0f46a8e6001be374200ae8a51fecb37f684/src/debugAdapter.ts#L9

> Why does disposing of the NamedPipeServerStream result in a timeout in the editor waiting for the response? Does it take too long? Is there a way to reconfigure the timeout to allow more time for clean up?

1. vscode tells inline DA "hey disconnect"
2. inline DA sends exact payload to the DA _server_ (this is via a named pipe, the creator of said pipe was the .NET server process)

> NOTE: DA server is written right now to clean up the `NamedPipeServerStream` on disconnect...

3. DA server cleans up NamedPipeServerStream as a part of handling the `disconnect` message thus disposing of the named pipe
4. _DA server fails silently_ to send the "hey I disconnected" response back to the Inline DA because the named pipe is already disposed of
5. VS Code times out waiting for "disconnect" response to come back from inline DA because inline DA never receives "disconnect" response from DA server

You've articulated the other side already:

1. vscode tells inline DA "hey disconnect"
2. inline DA sends exact payload to the DA _server_ (this is via a named pipe, the creator of said pipe was the .NET server process)
3. DA server sends a response to inline DA saying "I disconnected" and then starts to dispose of the named pipe and dispose of the DA server (these servers are created per-session and re-created after every session)
4. inline DA sends response to vscode - everyone is happy
5. For whatever reason, disposing of the NamedPipeServerStream is taking a very long time
6. User triggers a debug session again thus causing vscode to tell inline DA to start debugging
7. inline DA connects to named pipe file and sends first "initialize" message
8. NamedPipeServerStream is still not disposed for some reason and the "initialize" message is handled by the server that is about to be torn down.
9. NamedPipeServerStream gets disposed of finally and the DA server right after it... and a new DA server is created.
10. We are in a bad state because the initialize was handled by the previous DA server

These two scenarios are why I said this is the solution:

> Our DA server side shouldn't need to dispose of the named pipe and could potentially allow multiple connections over time (I have to see if this is configurable, otherwise I may have to handle the logic myself)
>
> I have some security concerns with this approach, but it might be the right idea (and is similar to how TCP-based DA's work today, I believe: have something open the whole time rather than recreating on disconnect)

in https://github.com/microsoft/debug-adapter-protocol/issues/126#issuecomment-655043170

> > My biggest wish would be that the disconnect request should really be 'fire and forget' from the vscode side to prevent DAP implementors from hitting this race condition.
>
> I do not think that would help.

You're right. That was before I said "In doing this exercise, I think I've answered my own question, but here it is." :)

**Comment by polinasok** [comment]
Thank you for the clarification @TylerLeonhardt

> VS Code times out waiting for "disconnect" response to come back from inline DA because inline DA never receives "disconnect" response from DA server

Since you control the inline DA, can't it use a different signal (such as detecting that the connection is lost, etc) to decide that it is time to send a disconnect response to the editor? What is the purpose of having an inline DA as opposed to communicating withe DA server directly?

**Comment by weinand** [comment]
If I understand Tyler correctly, the "inline DA" allows him to use named pipes for communicating with the DA (instead of a socket).
Today VS Code supports "named pipes" natively (thanks to a recent PR created by Tyler). With this the "inline DA" is no longer needed.

**Comment by weinand** [comment]
After reading the whole issue again, I now understand that the problem is about the **corner case** where the debug adapter and the debugger (together with its debuggee) are the same process. And I agree that the semantics of "disconnect" is a bit tricky in this case.

So let me try to clarify this:

The spec says:
> The ‘disconnect’ request is sent from the client to the debug adapter in order to stop debugging.
It asks the debug adapter to disconnect from the debuggee and to terminate the debug adapter.
If the debuggee has been started with the ‘launch’ request, the ‘disconnect’ request terminates the debuggee.
If the ‘attach’ request was used to connect to the debuggee, ‘disconnect’ does not terminate the debuggee.

From the perspective of a client (e.g. VS Code), the fact that DA and debugger/debuggee are the same process must not become visible. So the DA must not behave differently from setups where DA and debugger/debuggee are separate processes.

So, after the DA has received a ‘disconnect’ request from the client (e.g. VS Code), the DA must
- stop debugging the debuggee (by disconnecting the DA from the debugger/runtime),
- depending on the request type:
    - "launch":  terminate the debuggee (which typically means terminating the debugger/runtime)
    - "attach": leave the debuggee alone
- terminate itself (the DA).

When to send back the response for the ‘disconnect’ request is determined by how the combined DA/debugger works. If The DA continues to run after the debugger/debuggee has been terminated, the response can be sent late (and might include errors that occurred when terminating the debuggee). If the DA cannot survive the termination of the debugger/debuggee, then the response must be sent early, before terminating the debugger/debuggee.

This explanation was for the "single session mode" where a DA is launched for a single debug session.

If the DA is run in server node, the only difference is that the DA process is long running and there is no need for the "disconnect" request to terminate the DA process as the last step. Instead the communication channel with the client is closed.

@TylerLeonhardt in both setups I cannot really see the race condition.

**Comment by TylerLeonhardt** [comment]
Like I said, the fix is for the DA, which is in a server process, to actually run in "server mode" not in "single session mode". There's a race condition I [called out above](https://github.com/microsoft/debug-adapter-protocol/issues/126#issuecomment-761277342)... but it really only applies if you run a server process in "single session mode" which is silly to do to begin with.

I'm just gonna close this issue as I believe there's nothing the DAP needs to do about this.
